File: servers/Cour-Management/src/main/java/com/radim/project/entity/LessonContent.java
package com.radim.project.entity;

import com.radim.project.entity.enums.ContentType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "lesson_contents")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LessonContent {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "lesson_id", nullable = false)
    @ToString.Exclude
    private Lesson lesson;

    @Enumerated(EnumType.STRING)
    @NotNull
    private ContentType type;

    @Column(columnDefinition = "TEXT")
    private String textContent;

    private String pdfUrl;

    private String videoUrl;

    private String imageUrl;

    private UUID quizId; // Reference to a Quiz

    private int orderIndex;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: clients/mobile/lib/data/models/course/lesson_content.dart
class LessonContent {
  final String id;
  final String lessonId;
  final String type; // 'PDF' | 'TEXT' | 'VIDEO' | 'IMAGE' | 'QUIZ'
  final String? textContent;
  final String? pdfUrl;
  final String? videoUrl;
  final String? imageUrl;
  final String? quizId;
  final int orderIndex;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  LessonContent({
    required this.id,
    required this.lessonId,
    required this.type,
    this.textContent,
    this.pdfUrl,
    this.videoUrl,
    this.imageUrl,
    this.quizId,
    required this.orderIndex,
    this.createdAt,
    this.updatedAt,
  });

  factory LessonContent.fromJson(Map<String, dynamic> json) {
    return LessonContent(
      id: json['id'] as String,
      lessonId: json['lessonId'] as String,
      type: json['type'] as String,
      textContent: json['textContent'] as String?,
      pdfUrl: json['pdfUrl'] as String?,
      videoUrl: json['videoUrl'] as String?,
      imageUrl: json['imageUrl'] as String?,
      quizId: json['quizId'] as String?,
      orderIndex: json['orderIndex'] as int,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lessonId': lessonId,
      'type': type,
      if (textContent != null) 'textContent': textContent,
      if (pdfUrl != null) 'pdfUrl': pdfUrl,
      if (videoUrl != null) 'videoUrl': videoUrl,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (quizId != null) 'quizId': quizId,
      'orderIndex': orderIndex,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }
}

class CreateLessonContentRequest {
  final String type; // 'PDF' | 'TEXT' | 'VIDEO' | 'IMAGE' | 'QUIZ'
  final String? textContent;
  final String? pdfUrl;
  final String? videoUrl;
  final String? imageUrl;
  final String? quizId;
  final int orderIndex;

  CreateLessonContentRequest({
    required this.type,
    this.textContent,
    this.pdfUrl,
    this.videoUrl,
    this.imageUrl,
    this.quizId,
    required this.orderIndex,
  });

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      if (textContent != null) 'textContent': textContent,
      if (pdfUrl != null) 'pdfUrl': pdfUrl,
      if (videoUrl != null) 'videoUrl': videoUrl,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (quizId != null) 'quizId': quizId,
      'orderIndex': orderIndex,
    };
  }
}

